Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs

```diff
@@ -0,0 +1,228 @@
+			}
+			return ParseLeaf();
+		}
+		/// <summary>
+		/// Handles sums and substraction, it follows PEMDAS by asking if their left hand side and right hand side are multiplication
+		/// </summary>
+		/// <returns></returns>
+		Node ParseAddSubstract()
+		{
+			Node lhs = ParseMultiplyDivide();
+
+			while (true)
+			{
+				Func<int, int, int> op = null;
+				switch (_tokenizer.CurrentToken)
+				{
+					case Token.Sum:
+						op = (a, b) => a + b;
+						break;
+					case Token.Sub:
+						op = (a, b) => a - b;
+						break;
+				}
+				if (op == null)
+				{
+					return lhs;
+				}
+				_tokenizer.NextToken();
+
+				Node rhs = ParseMultiplyDivide();
+				lhs = new BinaryOperationNode(lhs, rhs, op);
+			}
+
+		}
+		/// <summary>
+		/// Resolves Multiplication and division in PEMDAS order, let <see cref="ParseUnary"/> resolve negatives and such expresions
+		/// </summary>
+		/// <returns></returns>
+		Node ParseMultiplyDivide()
+		{
+			Node lhs = ParseUnary();
+
+			while (true)
+			{
+				Func<int, int, int> op = null;
+				switch (_tokenizer.CurrentToken)
+				{
+					case Token.Mul:
+						op = (a, b) => a * b;
+						break;
+					case Token.Div:
+						op = (a, b) => a / b;
+						break;
+				}
+				if (op == null)
+				{
+					return lhs;
+				}
+				_tokenizer.NextToken();
+
+				Node rhs = ParseUnary();
+				lhs = new BinaryOperationNode(lhs, rhs, op);
+			}
+		}
+		/// <summary>
+		/// This recognize our terminal elements, like number and constants, also reduce parenthesis expresions
+		/// </summary>
+		/// <returns>Terminal element</returns>
+		Node ParseLeaf()
+		{
+			switch (_tokenizer.CurrentToken)
+			{
+				case Token.Number:
+					_tokenizer.NextToken();
+					return new NumberNode(_tokenizer.Number);
+				case Token.OpenParenthesis:
+					_tokenizer.NextToken();
+					Node node = ParseAddSubstract();
+
+					if (_tokenizer.CurrentToken != Token.CloseParenthesis)
+					{
+						throw new Exception("Missing Close Parenthesis");
+					}
+					_tokenizer.NextToken();
+
+					return node;
+				case Token.ConstantBottom:
+					_tokenizer.NextToken();
+					return new NumberNode(_rawSprite.RawTexture.Height / _rawSprite.FramesV);
+				case Token.ConstantTop:
+					_tokenizer.NextToken();
+					return new NumberNode(0);
+				case Token.ConstantRight:
+					_tokenizer.NextToken();
+					return new NumberNode(_rawSprite.RawTexture.Width / _rawSprite.FramesH);
+				case Token.ConstantLeft:
+					_tokenizer.NextToken();
+					return new NumberNode(0);
+				case Token.ConstantCenter:
+					_tokenizer.NextToken();
+					if (_yAxis)
+					{
+						return new NumberNode(_rawSprite.RawTexture.Height / _rawSprite.FramesV / 2);
+					}
+					return new NumberNode(_rawSprite.RawTexture.Width / _rawSprite.FramesH / 2);
+				default:
+					throw new Exception($"Unexpect token: {_tokenizer.CurrentToken}"); // TODO: Make a SyntaxException or something similar
+			}
+		}
+
+	}
+	#region NodeClasses
+
+	abstract class Node
```

Comment:
Why are all those in a single file?

Reply:
oh my, forgot to let refactor that when cleaning